tests/helpers/zos_lineinfile_helper.py

```diff
@@ -159,3 +186,126 @@ def DsGeneralResultKeyMatchesRegex(test_name, ansible_zos_module, test_env, test
         for key in kwargs:
             assert re.match(kwargs.get(key), result.get(key))
     clean_ds_test_env(test_env["DS_NAME"], hosts)
+
+
+def DsGeneralForce(ansible_zos_module, test_env, test_text, test_info, expected):
+    MEMBER_1, MEMBER_2 = "MEM1", "MEM2"
+    hosts = ansible_zos_module
+    test_info["path"] = DEFAULT_DATA_SET_NAME+"({0})".format(MEMBER_2)
+    if test_env["ENCODING"]:
+        test_info["encoding"] = test_env["ENCODING"]
+    try:
+        # set up:
+        # create pdse
+        hosts.all.zos_data_set(name=DEFAULT_DATA_SET_NAME, state="present", type="pdse", replace=True)
+        # add members
+        hosts.all.zos_data_set(
+            batch=[
+                {
+                    "name": DEFAULT_DATA_SET_NAME + "({0})".format(MEMBER_1),
+                    "type": "member",
+                    "state": "present",
+                    "replace": True,
+                },
+                {
+                    "name": DEFAULT_DATA_SET_NAME + "({0})".format(MEMBER_2),
+                    "type": "member",
+                    "state": "present",
+                    "replace": True,
+                },
+            ]
+        )
+        # write memeber to verify cases
+        hosts.all.shell(cmd="echo \"{0}\" > {1}".format(test_text, test_info["path"]))
+        # copy/compile c program and copy jcl to hold data set lock for n seconds in background(&)
+        hosts.all.zos_copy(content=c_pgm, dest='/tmp/disp_shr/pdse-lock.c', force=True)
+        hosts.all.zos_copy(
+            content=call_c_jcl.format(DEFAULT_DATA_SET_NAME, MEMBER_1),
+            dest='/tmp/disp_shr/call_c_pgm.jcl',
+            force=True
+        )
+        hosts.all.shell(cmd="xlc -o pdse-lock pdse-lock.c", chdir="/tmp/disp_shr/")
+
+        # submit jcl
+        hosts.all.shell(cmd="submit call_c_pgm.jcl", chdir="/tmp/disp_shr/")
+
+        # pause to ensure c code acquires lock
+        time.sleep(5)  
+        # call line infile to see results      
+        results = hosts.all.zos_lineinfile(**test_info)
+        pprint(vars(results))
+        if test_env["ENCODING"] == 'IBM-1047':
+            cmdStr =r"""cat "//'{0}'" """.format(test_info["path"])
+            results = hosts.all.shell(cmd=cmdStr)
+            pprint(vars(results))
+            for result in results.contacted.values():
+                print()
+                assert result.get("stdout").replace('\n', '').replace(' ', '') == expected.replace('\n', '').replace(' ', '')
```

Comment:
what happens if `test_env["ENCODING"] != 'IBM-1047'`? i don't see an else statement to catch that -- is it supposed to be an automatic success? Just want to be cautious of false "passing" test scenarios

Reply:
This is a good point, its marked resolved but see no code that addresses this. Although not that familiar with the test cases for this module it looks like there is a check for encodings that are not equal to 1047, that is done in [set_ds_test_env](https://github.com/ansible-collections/ibm_zos_core/blob/enhancement/356/Support_disposition_share_DISP%3DSHR/tests/helpers/zos_lineinfile_helper.py#L140), it appears this function is used to normalize the target you are about to test to get it to an particular encoding. There is a list of 3 encodings ([ENCODING = ['IBM-1047', 'ISO8859-1', 'UTF-8']](https://github.com/ansible-collections/ibm_zos_core/blob/ad0506cd66563c36c3c4193f373842f7b16eb474/tests/functional/modules/test_zos_lineinfile_func.py#L68)) that are often used but seem to have been skipped in the more recent tests cases added. 

I would say all these test cases need to be rewritten as they are difficult to follow but that is not in plan, but for now this is a valid question that was resolved and I don't see any response or code change to support the resolved comment. 

I can understand encodings being new to folks, but we need to ensure we are doing our testing adequately

Reply:
I didn't understand well the problem sorry just avoid the change of encoding but I work on it and I realize somethings:

Well I use the functions from other helpers but like here it doesn't check output to expected if it's another encoding too and is the function use for DS types [here](https://github.com/ansible-collections/ibm_zos_core/blob/ad0506cd66563c36c3c4193f373842f7b16eb474/tests/helpers/zos_lineinfile_helper.py#L138 ) also the other function call to make the assertion in case for another encoding just base with these `assert int(result.get("stdout")) != 0`. Could you explain me about the [assertion](https://github.com/ansible-collections/ibm_zos_core/blob/ad0506cd66563c36c3c4193f373842f7b16eb474/tests/helpers/zos_lineinfile_helper.py#L124)?

I'm working on the false passing cases for all other encodings.

Reply:
Unfortunately this test code is written in such a way its quite difficult to follow, from what i see it looks like ` set_ds_test_env` will check if the from encoding is ibm-1047, when it's not it ,it uses the encoding module to change the local data to match the too encoding. 

Eg, if the list of encodings being traversed lands on ISO-88591, it changes the content meaning it writes it out as ISO-88591 so that way it when it copies it to the target data set, the blockinfile/linefile tests can operate on the data in the encoding that matches. 

A data set can not contain ISO-88591 and then the blockinfile module think its IBM-1047 else it will break, no regex will work. 

So i think this function sets up the data on the target for the ansible module so that it does not break. To answer your questions now , `assert int(result.get("stdout")) != 0` , after all this copying and encoding conversions are done, the shell module is going to cat the data set to make sure there is content in the data set, and thus stdout should not be be false, (using 0 is poor idea here). 

This `cat` that checks for data really is not enough , it just tells you a copy was successful, it does not reverse the encoding to validate the enconding was successful but that is ok for now, as I said these tests cases should all be rewritten one at a time an managed within the test function not spread out over a large dictionary and a bunch of helpers. 

I think we could live with the way its written now, but we are limiting our encoding testing, whats funny is the modules manage all the encoding yet we have all this added complexity , there are easier ways to do this. Spend a bit of time on it, if its starts to make sense lets work on it, if we need to talk that is fine, also Fernando knows this code fairly well.

Reply:
I got a problem with the design because with zos_encode could work but time not to well, try with 
`iconv -f {0} -t {1} "//'{2}'" > "//'{2}'" """.format(test_info["encoding"], 'IBM-1047', test_info["path"])` to past all text with IBM-1047 except for the line added to return to IBM-1047 but had a problem because it change the message and throw garbage 
<img width="1108" alt="Captura de pantalla 2023-04-19 a la(s) 15 43 11" src="https://user-images.githubusercontent.com/68956970/233224239-fb94b863-4ad6-4216-90cc-482bd75629dd.png">

Reply:
I already add to cover all cases, but not verify as I would like the encodings

Reply:
Understood, I am not sure without some hands on investigation but I think the issue is the source file encoding is not what is iconv is being told is correct. 

Eg, if for a file on your controller is encoding in UTF8 (for example) and you tell iconv -f ISO-8859-1 -t IBM-1047, its going to end up like garbage because it mapped a UTF8 file as ISO-8859-1 to IBM-1047, this is a guess but i wonder if what you had seen was just a source issue i describe.